src/hpr_max_copies.erl

```diff
@@ -0,0 +1,141 @@
+-module(hpr_max_copies).
+
+-export([
+    init/0,
+    check/2,
+    cleanup/1
+]).
+
+-define(ETS, hpr_max_copies_ets).
+-define(MAX_COPIES, max_copies).
+-define(MAX_TOO_LOW, max_copies_max_too_low).
+-define(CLEANUP_TIME, timer:hours(1)).
+
+-spec init() -> ok.
+init() ->
+    %% Table structure
+    %% {Key :: binary(), Counter :: non_neg_integer() :: Timestamp :: integer()}
+    ets:new(?ETS, [
+        public,
+        named_table,
+        set,
+        {write_concurrency, true}
+    ]),
+    ok = scheduled_cleanup(?CLEANUP_TIME),
+    ok.
+
+-spec check(Key :: binary(), Max :: non_neg_integer()) ->
+    ok | {error, ?MAX_TOO_LOW | ?MAX_COPIES}.
+check(_Key, Max) when Max =< 0 ->
+    {error, ?MAX_TOO_LOW};
+check(Key, Max) ->
+    case
+        ets:update_counter(
+            ?ETS, Key, {2, 1}, {default, 0, erlang:system_time(millisecond)}
+        )
+    of
+        C when C > Max ->
+            {error, ?MAX_COPIES};
+        _C ->
+            ok
+    end.
+
+-spec cleanup(Duration :: non_neg_integer()) -> ok.
+cleanup(Duration) ->
+    erlang:spawn(fun() ->
+        Time = erlang:system_time(millisecond) - Duration,
+        Deleted = ets:select_delete(?ETS, [
+            {{'_', '_', '$3'}, [{'<', '$3', Time}], [true]}
+        ]),
+        lager:debug("expiring ~w keys", [Deleted])
+    end),
+    ok.
+
+%% ------------------------------------------------------------------
+%% Internal Function Definitions
+%% ------------------------------------------------------------------
+
+-spec scheduled_cleanup(Duration :: non_neg_integer()) -> ok.
+scheduled_cleanup(Duration) ->
+    {ok, _} = timer:apply_interval(Duration, ?MODULE, cleanup, [Duration]),
```

Comment:
Minor concern is to avoid having more than one of these running at a time should it get stuck for some reason. But the interval times are quite long, so I think we're fine here.